Documentation/proposals/20250122-blackbox-crd.md

```diff
@@ -0,0 +1,148 @@
+
+## Goals
+
+- Provide a way for users to self-service adding module config for their Probe CRD
+- Consolidate the module configuration generation logic in a central point for other resources to use
+
+## Audience
+
+* Administrators providing Prometheus monitoring who want to enable self-service probe configuration for their users
+* Teams wanting to manage blackbox module configurations declaratively through Kubernetes CRDs
+* Users requiring a standardized, Kubernetes-native approach to monitor external targets
+
+## Non-Goals
+
+- This propose don't aim to replace the probe crd
+
+## How
+
+We propose creating two new Custom Resource Definitions (CRDs):
+- The BlackboxExporter CRD, which functions similarly to the Prometheus CRD, will manage the deployment configuration and lifecycle of blackbox-exporter instances
+- The BlackboxExporterModule CRD will define module specifications and automatically generate blackbox configurations, analogous to how Alertmanager works in Prometheus Operator
+
+## Alternatives
+
+- Update blackbox configuration whenever a probe module is modified or added
+
+## Action Plan
+
+1. Create the BlackboxExporter CRD which will manage blackbox-exporter instances as either a DaemonSet or Deployment
+2. Implement a configuration reloader sidecar container to automatically reload blackbox-exporter when configuration changes, similar to how Prometheus Operator handles configuration updates
+
+## CRDs Drafg
+
+```yaml
+apiVersion: monitoring.zhipu.ai/v1alpha1
+kind: BlackboxExporter
+metadata:
+  name: blackbox-exporter-draft
+spec:
+  # Deployment mode - how the BlackboxExporter should be deployed 
+    # Or Daemonset mode
+  deployMode: Deployment
+  replicas: 2
+  
+  # Key configuration options
+  baseImage: prom/blackbox-exporter
+  timeoutOffset: "0.5"
+  logLevelProber: "info"
+  
+  # Module selection - allows selecting which probe modules to include
+  moduleSelector:
+    matchLabels:
+      type: http-probe
+  
+  # Resource requirements
+  resources:
+    limits:
+      memory: 200Mi
+      cpu: 100m
+    requests:
+      memory: 100Mi
+      cpu: 50m
+      
+  # Advanced configuration options
+  securityContext:
+    runAsNonRoot: true
+    runAsUser: 1000
+  
+  # Network configuration
+  webListenAddresses:
+    - ":9115"
+  externalURL: "https://blackbox.example.com"
+  
+status:
+  # Status fields will be managed by the controller
+  replicas: 2
+  availableReplicas: 2
+  updatedReplicas: 2
+  unavailableReplicas: 0
+  conditions:
+    - type: Available
+      status: "True"
+      lastTransitionTime: "2025-02-28T12:00:00Z"
+      reason: MinimumReplicasAvailable
+      message: "All replicas are available"
+
+---
+apiVersion: monitoring.zhipu.ai/v1alpha1
+kind: BlackboxExporterModule
+metadata:
+  name: example-http-probe
+  labels:
+    type: http-probe
+spec:
+  # HTTP probe configuration
+  # Common probe settings
+  timeout: 5s
+  prober: http
+  http:
```

Comment:
This resembles the BB configuration as is, but I was thinking maybe we can add something like, WDYT?

```yaml
prober: http // (http, tcp, dns, icmp, grpc)
proberConfg:
    http:
      method: GET
      headers:
        Host: host.example.com
    icmp:
      preferred_ip_protocol: "ip4"
```

Reply:
Hi @Pger-Y you're right, I was thinking a bit about the current AlertmanagerConfig structure when I_ suggested this - https://doc.crds.dev/github.com/prometheus-operator/prometheus-operator/monitoring.coreos.com/AlertmanagerConfig/v1alpha1@v0.80.0. But we can just keep this simpler for a first version. But at the same time probably keeping as close as the current configuration could be a good idea, just in terms of structure, for context https://github.com/prometheus/blackbox_exporter/blob/master/example.yml

Reply:
@Pger-Y the API structure should follow the general Kube API conventions. While it may seem counter-intuitive to not always follow the Prometheus config format, the Kube conventions exist for good reasons.

https://prometheus-operator.dev/docs/community/contributing/#changes-to-the-apis